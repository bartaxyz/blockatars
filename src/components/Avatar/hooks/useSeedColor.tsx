import { useMemo } from "react";
import { useSeedRandom } from "./useSeedRandom";

interface UseSeedColorReturnType {
  backgroundColor: string;
  color1: string;
  color2: string;
}

export const useSeedColor = (seed: string): UseSeedColorReturnType => {
  const { shift } = useSeedRandom(seed);

  return useMemo(() => {
    const createColor = () => {
      // saturation is the whole color spectrum
      const h = Math.floor(shift() * 360) + "deg";

      // saturation goes from 40 to 100, it avoids greyish colors
      const s = (shift() * 60 + 40).toFixed(1) + "%";

      // lightness can be anything from 0 to 100, but probabilities are a bell curve around 50%
      const l = ((shift() + shift() + shift() + shift()) * 25).toFixed(1) + "%";

      return "hsl(" + h + ", " + s + ", " + l + ")";
    };

    return {
      backgroundColor: createColor(),
      color1: createColor(),
      color2: createColor(),
    };
  }, [seed]);
};
