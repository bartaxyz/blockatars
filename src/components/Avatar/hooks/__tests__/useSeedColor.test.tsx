// write jest test for useSeedColor hook

import { renderHook } from "@testing-library/react";
import { useSeedColor } from "../useSeedColor";

describe("useSeedColor", () => {
  test("returns a color", () => {
    const returnValue = renderHook(() =>
      useSeedColor("0xe6cc08e94d44184e5f412e52f9f1866b6d7c8990")
    );

    expect(returnValue.result.current).toMatchSnapshot();
  });
});
