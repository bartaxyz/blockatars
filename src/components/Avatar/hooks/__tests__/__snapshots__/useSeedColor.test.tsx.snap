// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useSeedColor returns a color 1`] = `
{
  "backgroundColor": "hsl(169deg, 80.2%, 60.7%)",
  "color1": "hsl(294deg, 87.8%, 60.5%)",
  "color2": "hsl(332deg, 76.9%, 50.6%)",
}
`;
